### ISSUE TITLE
How to checkpoint model and get val_acc and val_loss with train_on_batch() and test_on_batch() functions?

### ISSUE THREAD

**Issue by saanikagupta** [keras-team/keras#13061]
My dataset has 396 CSV files. I want to train on 297 files and test on the remaining 99 files. Since each CSV file has different number of rows, where each row is one data point and I want my batch-size to be the size of the CSV file. This means my batch-size has to vary with the input, so I used train_on_batch() and test_on_batch(). I am more familiar with fit_generator(), and it used to output acc, loss, val_acc and val_loss with verbose = 1. But in train_on_batch, there's no parameter verbose. So I hard coded the training loss and accuracy but not sure how to get the val_acc and val_loss while training, just like fit_generator(). Is this even possible? Can I get this for each step/batch? 

I also wanted my model to save weights after every 50 epochs. But train_on_batch() has no argument 'callbacks'. Also in the 'monitor' parameter of ModelCheckpoint(), it takes val_loss or val_acc as the input. But I am not sure how to get val_loss or val_acc from train_on_batch() as I told, so I don't understand how to proceed with this. 

Here is the relevant code snippet:

    model.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics = ['mae'])
    # checkpoint
    filepath = "weights.best.hdf5"
    checkpoint = ModelCheckpoint(filepath, monitor = 'val_acc', verbose = 1, save_best_only = 
    True, mode = 'max')
    callbacks_list = [checkpoint]
    print("Train on {} samples" .format(len(train_csv_list)))
    for epoch in range(max_epoch):
      print('Epoch %d/%d' % (epoch + 1, max_epoch))
      tic = time.time()
      i = 1
      for CSV in train_csv_list:
          df = pd.read_csv(CSV, header = 0)
          x_train = df.iloc[:, 1: -1] # dropping 1st and last column
          y_train = df.iloc[:, -1:]
          x_train_goal_information = x_train.iloc[:, -2:]
          x_train = x_train.iloc[:, :-2]

          # Changing to 3-dim for Conv1D
          x_train = np.expand_dims(x_train, axis = 2)
          # print ("x_train shape: " + str(x_train.shape))
    
          [loss, acc] = model.train_on_batch([x_train, x_train_goal_information], y_train)
          info = 'Step: {0:d} - loss: {1:6.10f} - acc: {2:6.10f}'
          print(info.format(i, loss, acc))
          i = i + 1
    toc = time.time()
    print('Elapsed time:', toc - tic, 'seconds')

Please give me a way to get all the validation losses and accuracies and also how can I save my model weights after every 50 epochs so that later I can load these saved weights and further train the model.

**Comment by dabasajay** [comment]
Hello! `train_on_batch` function only takes 4 parameters namely `X`,Y`,`sample_weight` and `class_weight`. `fit_generator` internally uses `train_on_batch` function to run a single gradient update on a single batch of data and uses callbacks to do rest of the fancy stuff like a checkpoint, progress bar, etc. So you really can't use `train_on_batch` function to achieve what you want. Since you don't want to train on batches, just use `fit` function. For any reason, if you still want to use `fit_generator` then set steps_per_epoch=1 and inside data_generator, yield all of the data at once.
